### PR TITLE
First Round Tests and Refactors for CAS Login

### PR DESCRIPTION
**Pull request by StetsenTech**
**Ticket(s)**:
https://app.asana.com/0/1133543009734854/1150084536382593/f

**Description**:
This is the first round of tests and refactors to the login route. Now that I have been exposed to Rails and done some reading, I understand what needs to be done. This adds environment application configs, adds missing db columns for RubyCAS Core ActiveRecord gem, cleans up Double Render errors, removes invalid code, and some styling changes. 

**Summary**:
- Adds environment based configs for RubyCAS YML
- Updates migration schema
- Refactors some logic from the original brought over code
- Brings over BeyondZ authenticator since we will be using the system for a bit
- Removes code that is not longer valid in the transfer Sinatra to Rails
- Adds tests for `/login`

### CODE REVIEW

File: app/controllers/cas_controller.rb [hunk 1]
```diff
@@ -46,55 +50,57 @@ def login
...
     if params.has_key? 'onlyLoginForm'
       if request.env['HTTP_HOST']
-        guessed_login_uri = "#{request.env['HTTPS']}://#{request.env['REQUEST_URI']}#{self / '/login'}"
+        @protocol = request.env['HTTPS'] == 'on' ? 'https' : 'http'
+        guessed_login_uri = "#{@protocol}://#{request.env['REQUEST_URI']}/login"
       else
         guessed_login_uri = nil
       end
 
       @form_action = params['submitToURI'] || guessed_login_uri
```
Comment (StetsenTech) [inline app/controllers/cas_controller.rb]: I'm thinking of removing this logic for guessing a form action if they `onlyLoginForm`. If someone specifies onlyLoginForm I would prefer if they also specify `submitToURI`. I never want to be in a situation of trying to guess what URL someone else wants (especially for external systems).

File: app/controllers/cas_controller.rb [hunk 2]
```diff
@@ -103,42 +109,33 @@ def login
...
+    c = request.env['HTTP_X_FORWARDED_FOR'] || request.env['REMOTE_HOST'] || request.env['REMOTE_ADDR']
 
-    if error = validate_login_ticket(@lt)
+    if !result = validate_login_ticket(@lt)
       @message = {:type => 'mistake', :message => error}
       # generate another login ticket to allow for re-submitting the form
-      c = request.env['HTTP_X_FORWARDED_FOR'] || request.env['REMOTE_HOST'] || request.env['REMOTE_ADDR']
       @lt = LT.generate_login_ticket(c).ticket
-      return render :login, status: :internal_server_error
+      return render :login, status: :unauthorized
```
Comment (StetsenTech) [inline app/controllers/cas_controller.rb]: I hate returning 500s in general. If you know there is a problem and purposely doing a return, then it needs to be an error in the 400s.

File: app/controllers/cas_controller.rb [hunk 3]
```diff
@@ -258,25 +252,14 @@ def logout
...
     elsif @continue_url
-      render :logout
-    else
-      render :login
+      return render :logout
     end
+    render :login
   end
 
   def loginTicket
```
Comment (StetsenTech) [inline app/controllers/cas_controller.rb]: Anything below this point in this file is going to be in another PR. I made these changes so I don't forget when I start working on tests for these.

File: config/rubycas.yml [hunk 4]
```diff
@@ -1,26 +1,76 @@
...
+  cookie_domain: .localhost 
+  organization: "RSPEC-TEST"
+  infoline: "This is an rspec test."  
+  theme: simple
+  database:
+    <<: *default_db
+    database: platform_test
+  authenticator:
+    <<: *default_bz_auth
+    server: joinweb
```
Comment (StetsenTech) [inline config/rubycas.yml]: I may point this to `platform_test` instead...or I may just make another authenticator. 🤔

File: lib/beyondz.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+      request = Net::HTTP::Post.new('/users/check_credentials')
+      request.set_form_data(
+        'username' => credentials[:username],
+        'password' => credentials[:password]
+      )
+      response = http.request(request)
+      return (response.body == "true")
+    end
+  end
+end
```
Comment (StetsenTech) [inline lib/beyondz.rb]: Brought directly from SSO and the class it inherits from.

File: app/controllers/cas_controller.rb (hunk 1 above)
Comment (rshipp) [reply app/controllers/cas_controller.rb]: I think that'd be a good change, if we're sure it won't break anything.

File: app/controllers/cas_controller.rb [hunk 6]
```diff
@@ -103,42 +109,33 @@ def login
...
     @password = params['password']
     @lt = params['lt']
 
-    # Remove leading and trailing widespace from username.
-    @username.strip! if @username
-
-    if @username && settings.config[:downcase_username]
-      logger.debug("Converting username #{@username.inspect} to lowercase because 'downcase_username' option is enabled.")
-      @username.downcase!
-    end
```
Comment (rshipp) [inline app/controllers/cas_controller.rb]: thanks for cleaning up this logic

File: app/controllers/cas_controller.rb (hunk 2 above)
Comment (rshipp) [reply app/controllers/cas_controller.rb]: will this only happen when the credentials were actually wrong? if it could be an upstream communication error with join or something, I think it makes sense to have a 500

File: app/controllers/cas_controller.rb [hunk 7]
```diff
@@ -191,16 +183,17 @@ def loginpost
...
       end
-    rescue Core::Authenticator::AuthenticatorError => e
+    rescue RubyCAS::Server::Core::AuthenticatorError => e
       logger.error(e)
       # generate another login ticket to allow for re-submitting the form
-      c = request.env['HTTP_X_FORWARDED_FOR'] || request.env['REMOTE_HOST'] || request.env['REMOTE_ADDR']
       @lt = LT.generate_login_ticket(c).ticket
       @message = {:type => 'mistake', :message => e.to_s}
-      render :json => @message, status: :unauthorized
+      return render :login , status: :unauthorized
```
Comment (rshipp) [inline app/controllers/cas_controller.rb]: are we no longer using `@message` here?

File: config/rubycas.yml [hunk 8]
```diff
@@ -1,26 +1,76 @@
...
 default: &default
-  log:
-    output: log/casserver.log
-    level: INFO
   maximum_unused_login_ticket_lifetime: 300
   maximum_unused_service_ticket_lifetime: 300
   maximum_session_lifetime: 172800
   maximum_remember_me_lifetime: 604800
-  
+  infoline: Powered by <a href="http://rubycas.github.com">RubyCAS-Server-Core</a>
```
Comment (rshipp) [inline config/rubycas.yml]: do we need this? it's just extra fingerprinting info

**Comment by rshipp** [comment]
Looks like the tests are complaining because they don't know what `SSO_URL` to use

File: app/controllers/cas_controller.rb (hunk 7 above)
Comment (StetsenTech) [reply app/controllers/cas_controller.rb]: `@message` and most variable available in scope of the route are passed to render when using an ERB view.

File: config/rubycas.yml (hunk 8 above)
Comment (StetsenTech) [reply config/rubycas.yml]: Don't really need it. It's pretty common practice say what powers your system. I left it in because it is used in the login route. 

I think I am going to strip out some things in the view pages so we can have a fresh start.

File: app/controllers/cas_controller.rb (hunk 7 above)
Comment (rshipp) [reply app/controllers/cas_controller.rb]: ahhh ok, didn't know that

File: config/rubycas.yml (hunk 8 above)
Comment (rshipp) [reply config/rubycas.yml]: oh does this show up on the login page or something? I was thinking it was just a version banner.

File: app/controllers/cas_controller.rb (hunk 7 above)
Comment (StetsenTech) [reply app/controllers/cas_controller.rb]: To clarify and be more specific, the variables with `@` will get passed in.

File: app/controllers/cas_controller.rb (hunk 2 above)
Comment (StetsenTech) [reply app/controllers/cas_controller.rb]: This particular path only happens if you fail authentication.

**Comment by StetsenTech** [comment]
I'm going to go ahead and merge this into the feature branch. Feel free to still leave comments and I will make the updates in a future PR.
